<a href="https://colab.research.google.com/github/jaiswalakanksha22-glitch/Python-AI/blob/main/Project3_Jaiswal_Akanksha.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Steps and Questions**
Load the dataset, PRSA_Data.csv, into memory.

In [1]:
from google.colab import files
uploaded = files.upload()

Saving PRSA_Data_Aotizhongxin_20130301-20170228.csv to PRSA_Data_Aotizhongxin_20130301-20170228.csv
Saving PRSA_Data_Changping_20130301-20170228.csv to PRSA_Data_Changping_20130301-20170228.csv
Saving PRSA_Data_Dingling_20130301-20170228.csv to PRSA_Data_Dingling_20130301-20170228.csv
Saving PRSA_Data_Dongsi_20130301-20170228.csv to PRSA_Data_Dongsi_20130301-20170228.csv
Saving PRSA_Data_Guanyuan_20130301-20170228.csv to PRSA_Data_Guanyuan_20130301-20170228.csv
Saving PRSA_Data_Gucheng_20130301-20170228.csv to PRSA_Data_Gucheng_20130301-20170228.csv
Saving PRSA_Data_Huairou_20130301-20170228.csv to PRSA_Data_Huairou_20130301-20170228.csv
Saving PRSA_Data_Nongzhanguan_20130301-20170228.csv to PRSA_Data_Nongzhanguan_20130301-20170228.csv
Saving PRSA_Data_Shunyi_20130301-20170228.csv to PRSA_Data_Shunyi_20130301-20170228.csv
Saving PRSA_Data_Tiantan_20130301-20170228.csv to PRSA_Data_Tiantan_20130301-20170228.csv
Saving PRSA_Data_Wanliu_20130301-20170228.csv to PRSA_Data_Wanliu_20130301-2

In [15]:
import pandas as pd
import glob

files = glob.glob('/content/PRSA_Data_*.csv')
print(files)

df_list = [pd.read_csv(file) for file in files]
df = pd.concat(df_list, ignore_index=True)

df.head()

['/content/PRSA_Data_Aotizhongxin_20130301-20170228.csv', '/content/PRSA_Data_Gucheng_20130301-20170228.csv', '/content/PRSA_Data_Wanshouxigong_20130301-20170228.csv', '/content/PRSA_Data_Shunyi_20130301-20170228.csv', '/content/PRSA_Data_Tiantan_20130301-20170228.csv', '/content/PRSA_Data_Nongzhanguan_20130301-20170228.csv', '/content/PRSA_Data_Changping_20130301-20170228.csv', '/content/PRSA_Data_Guanyuan_20130301-20170228.csv', '/content/PRSA_Data_Huairou_20130301-20170228.csv', '/content/PRSA_Data_Dingling_20130301-20170228.csv', '/content/PRSA_Data_Dongsi_20130301-20170228.csv', '/content/PRSA_Data_Wanliu_20130301-20170228.csv']


,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,4.0,4.0,4.0,7.0,300.0,77.0,-0.7,1023.0,-18.8,0.0,NNW,4.4,Aotizhongxin
1,2,2013,3,1,1,8.0,8.0,4.0,7.0,300.0,77.0,-1.1,1023.2,-18.2,0.0,N,4.7,Aotizhongxin
2,3,2013,3,1,2,7.0,7.0,5.0,10.0,300.0,73.0,-1.1,1023.5,-18.2,0.0,NNW,5.6,Aotizhongxin
3,4,2013,3,1,3,6.0,6.0,11.0,11.0,300.0,72.0,-1.4,1024.5,-19.4,0.0,NW,3.1,Aotizhongxin
4,5,2013,3,1,4,3.0,3.0,12.0,12.0,300.0,72.0,-2.0,1025.2,-19.5,0.0,N,2.0,Aotizhongxin


In [4]:
print(df.shape)
print(df.columns)

(420768, 18)
Index(['No', 'year', 'month', 'day', 'hour', 'PM2.5', 'PM10', 'SO2', 'NO2',
       'CO', 'O3', 'TEMP', 'PRES', 'DEWP', 'RAIN', 'wd', 'WSPM', 'station'],
      dtype='object')


Clean the data and check missing values for this dataset.

In [16]:
df.dtypes

,0
No,int64
year,int64
month,int64
day,int64
hour,int64
PM2.5,float64
PM10,float64
SO2,float64
NO2,float64
CO,float64


In [17]:
# Check missing values column-wise
df.isnull().sum(axis=0)


,0
No,0
year,0
month,0
day,0
hour,0
PM2.5,8739
PM10,6449
SO2,9021
NO2,12116
CO,20701


In [18]:
# Drop unnecessary column
df = df.drop(columns=['No'])

# Interpolate numerical columns
num_cols = df.select_dtypes(include=['float64', 'int64']).columns
df[num_cols] = df[num_cols].interpolate(method='linear')

# Forward fill categorical column
df['wd'] = df['wd'].ffill()

# Backward fill remaining edge NaNs
df = df.bfill()

# Final check
print(df.isnull().sum())

year       0
month      0
day        0
hour       0
PM2.5      0
PM10       0
SO2        0
NO2        0
CO         0
O3         0
TEMP       0
PRES       0
DEWP       0
RAIN       0
wd         0
WSPM       0
station    0
dtype: int64


In [19]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 420768 entries, 0 to 420767
Data columns (total 17 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   year     420768 non-null  int64  
 1   month    420768 non-null  int64  
 2   day      420768 non-null  int64  
 3   hour     420768 non-null  int64  
 4   PM2.5    420768 non-null  float64
 5   PM10     420768 non-null  float64
 6   SO2      420768 non-null  float64
 7   NO2      420768 non-null  float64
 8   CO       420768 non-null  float64
 9   O3       420768 non-null  float64
 10  TEMP     420768 non-null  float64
 11  PRES     420768 non-null  float64
 12  DEWP     420768 non-null  float64
 13  RAIN     420768 non-null  float64
 14  wd       420768 non-null  object 
 15  WSPM     420768 non-null  float64
 16  station  420768 non-null  object 
dtypes: float64(11), int64(4), object(2)
memory usage: 54.6+ MB


,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,WSPM
count,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000,420768.000000
mean,2014.662560,6.522930,15.729637,11.500000,79.839718,104.910268,15.913090,50.600836,1235.682649,57.237872,13.531692,1010.753337,2.482421,0.064428,1.730034
std,1.177198,3.448707,8.800102,6.922195,80.950217,92.431422,21.896609,35.170775,1161.790893,57.135195,11.437867,10.474302,13.797675,0.820638,1.246674
min,2013.000000,1.000000,1.000000,0.000000,2.000000,2.000000,0.285600,1.026500,100.000000,0.214200,-19.900000,982.400000,-43.400000,0.000000,0.000000
25%,2014.000000,4.000000,8.000000,5.750000,20.000000,36.000000,3.000000,23.000000,500.000000,10.000000,3.100000,1002.300000,-8.900000,0.000000,0.900000
50%,2015.000000,7.000000,16.000000,11.500000,55.000000,82.000000,7.000000,43.000000,900.000000,44.000000,14.500000,1010.400000,3.000000,0.000000,1.400000
75%,2016.000000,10.000000,23.000000,17.250000,111.000000,145.000000,20.000000,71.000000,1500.000000,82.000000,23.300000,1019.000000,15.100000,0.000000,2.200000
max,2017.000000,12.000000,31.000000,23.000000,999.000000,999.000000,500.000000,290.000000,10000.000000,1071.000000,41.600000,1042.800000,29.100000,72.500000,13.200000


Perform feature engineering on the date feature.

In [20]:
df['datetime'] = pd.to_datetime(df[['year', 'month', 'day', 'hour']])
df = df.sort_values(by='datetime')

In [21]:
# Basic time features
df['day_of_week'] = df['datetime'].dt.dayofweek   # 0=Monday
df['week_of_year'] = df['datetime'].dt.isocalendar().week
df['quarter'] = df['datetime'].dt.quarter

# Weekend indicator
df['is_weekend'] = df['day_of_week'].isin([5, 6]).astype(int)

In [22]:
# Hour encoding
import numpy as np

df['hour_sin'] = np.sin(2 * np.pi * df['hour'] / 24)
df['hour_cos'] = np.cos(2 * np.pi * df['hour'] / 24)

# Month encoding
df['month_sin'] = np.sin(2 * np.pi * df['month'] / 12)
df['month_cos'] = np.cos(2 * np.pi * df['month'] / 12)

# Day encoding
df['day_sin'] = np.sin(2 * np.pi * df['day'] / 31)
df['day_cos'] = np.cos(2 * np.pi * df['day'] / 31)

In [23]:
# Drop Redundant Columns
df = df.drop(columns=['year', 'month', 'day', 'hour'])

Split the data into 80% of training and 20% of the test dataset.

In [31]:
df = df.sort_values(by='datetime')

split_index = int(len(df)*0.8)

train_df = df.iloc[:split_index]
test_df = df.iloc[split_index:]

X_train = train_df.drop(columns=['wd','datetime'])
y_train = train_df['wd']

X_test = test_df.drop(columns=['wd','datetime'])
y_test = test_df['wd']

Preprocess the data using the normalization method to convert all features into the range of [0,1]

In [33]:
# Drop or encode categorical
X_train = X_train.drop(columns=['station'], errors='ignore')
X_test = X_test.drop(columns=['station'], errors='ignore')

# Scale
from sklearn.preprocessing import MinMaxScaler, LabelEncoder
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Encode target
le = LabelEncoder()
y_train_enc = le.fit_transform(y_train)
y_test_enc = le.transform(y_test)


Build a neuron network with two hidden layers of 20 and 10 neurons to forecast wd using all other features and TensorFlow. Does it overfit or underfit the data? Please justify your answer.

In [36]:
import tensorflow as tf
from tensorflow.keras import layers

model = tf.keras.Sequential([
    layers.Dense(20, activation='relu', input_shape=(X_train_scaled.shape[1],)),
    layers.Dense(10, activation='relu'),
    layers.Dense(len(le.classes_), activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(X_train_scaled, y_train_enc,
                    validation_data=(X_test_scaled, y_test_enc),
                    epochs=50, batch_size=32)

Epoch 1/50
10520/10520 ━━━━━━━━━━━━━━━━━━━━ 26s 2ms/step - accuracy: 0.1446 - loss: 2.5556 - val_accuracy: 0.1699 - val_loss: 2.5157
Epoch 2/50
10520/10520 ━━━━━━━━━━━━━━━━━━━━ 26s 2ms/step - accuracy: 0.1585 - loss: 2.4998 - val_accuracy: 0.1604 - val_loss: 2.4976
Epoch 3/50
10520/10520 ━━━━━━━━━━━━━━━━━━━━ 26s 2ms/step - accuracy: 0.1638 - loss: 2.4816 - val_accuracy: 0.1602 - val_loss: 2.4864
Epoch 4/50
10520/10520 ━━━━━━━━━━━━━━━━━━━━ 25s 2ms/step - accuracy: 0.1666 - loss: 2.4729 - val_accuracy: 0.1639 - val_loss: 2.4800
Epoch 5/50
10520/10520 ━━━━━━━━━━━━━━━━━━━━ 25s 2ms/step - accuracy: 0.1693 - loss: 2.4671 - val_accuracy: 0.1776 - val_loss: 2.4691
Epoch 6/50
10520/10520 ━━━━━━━━━━━━━━━━━━━━ 41s 2ms/step - accuracy: 0.1699 - loss: 2.4626 - val_accuracy: 0.1728 - val_loss: 2.4667
Epoch 7/50
10520/10520 ━━━━━━━━━━━━━━━━━━━━ 25s 2ms/step - accuracy: 0.1708 - loss: 2.4582 - val_accuracy: 0.1702 - val_loss: 2.4633
Epoch 8/50
10520/10520 ━━━━━━━━━━━━━━━━━━━━ 25s 2ms/step - accuracy: 

In [39]:
train_loss, train_mae = model.evaluate(X_train_scaled, y_train_enc, verbose=0)
test_loss, test_mae = model.evaluate(X_test_scaled, y_test_enc, verbose=0)

print("Training Loss (MSE):", train_loss)
print("Training MAE:", train_mae)
print("Testing Loss (MSE):", test_loss)
print("Testing MAE:", test_mae)

Training Loss (MSE): 2.415334701538086
Training MAE: 0.1814303696155548
Testing Loss (MSE): 2.444202184677124
Testing MAE: 0.17188724875450134


**Overfitting or Underfitting? (TensorFlow)**
The model is underfitting because both training and validation accuracies are low (~18%) and very close to each other. This indicates that the model is unable to learn meaningful patterns from the data. Additionally, the loss decreases very slowly and remains high, suggesting insufficient model capacity or feature representation. Since there is no significant gap between training and validation performance, the model is not overfitting but rather underfitting.

Tune the model using the following hyperparameters using keras-tuner:
First hidden layer with units between 20 and 100 with a step size of 5
Second hidden layer with units between 10 and 50 with a step size of 2
The dropout rate for both hidden layer is between 0.2 and 0.8 with a step size of 0.1

In [41]:
!pip install -q keras-tuner

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.4/129.4 kB 2.0 MB/s eta 0:00:00


In [42]:
import keras_tuner as kt

def build_model(hp):
    model = tf.keras.Sequential()

    # First layer
    units1 = hp.Int('units1', 20, 100, step=5)
    dropout1 = hp.Float('dropout1', 0.2, 0.8, step=0.1)

    model.add(layers.Dense(units1, activation='relu'))
    model.add(layers.Dropout(dropout1))

    # Second layer
    units2 = hp.Int('units2', 10, 50, step=2)
    dropout2 = hp.Float('dropout2', 0.2, 0.8, step=0.1)

    model.add(layers.Dense(units2, activation='relu'))
    model.add(layers.Dropout(dropout2))

    model.add(layers.Dense(len(le.classes_), activation='softmax'))

    model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])

    return model

tuner = kt.RandomSearch(
    build_model,
    objective='val_accuracy',
    max_trials=10,
    directory='tuner_dir'
)

tuner.search(X_train_scaled, y_train_enc,
             validation_data=(X_test_scaled, y_test_enc),
             epochs=20)

Trial 10 Complete [00h 10m 07s]
val_accuracy: 0.18896308541297913

Best val_accuracy So Far: 0.18950970470905304
Total elapsed time: 01h 40m 41s


Please recommend the optimal model and specify the hyperparameters.



In [43]:
best_hp = tuner.get_best_hyperparameters(num_trials=1)[0]

print("Best Hyperparameters:")
print("Units Layer 1:", best_hp.get('units1'))
print("Units Layer 2:", best_hp.get('units2'))
print("Dropout Layer 1:", best_hp.get('dropout1'))
print("Dropout Layer 2:", best_hp.get('dropout2'))

Best Hyperparameters:
Units Layer 1: 20
Units Layer 2: 16
Dropout Layer 1: 0.30000000000000004
Dropout Layer 2: 0.4


The model still underfits the data as both training and validation accuracies remain low (~19%), indicating limited predictive capability of the features for wind direction. Tree-based models like Random Forest may perform better for this dataset compared to neural networks.

Build a neuron network with two hidden layers of 20 and 10 neurons to forecast wd using all other features and **PyTorch**. Does it overfit or underfit the data? Please justify your answer.

In [46]:
import torch
import torch.nn as nn
import torch.optim as optim

# Define model
class WindModel(nn.Module):
    def __init__(self, input_size, output_size):
        super(WindModel, self).__init__()

        self.model = nn.Sequential(
            nn.Linear(input_size, 20),
            nn.ReLU(),

            nn.Linear(20, 10),
            nn.ReLU(),

            nn.Linear(10, output_size)
        )

    def forward(self, x):
        return self.model(x)

# Initialize model
input_size = X_train_scaled.shape[1]
output_size = len(le.classes_)

model = WindModel(input_size, output_size)

In [47]:
# Prepare data
X_train_tensor = torch.tensor(X_train_scaled, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train_enc, dtype=torch.long)

X_test_tensor = torch.tensor(X_test_scaled, dtype=torch.float32)
y_test_tensor = torch.tensor(y_test_enc, dtype=torch.long)

In [48]:
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

train_losses = []
test_losses = []

for epoch in range(50):
    # Training
    model.train()
    optimizer.zero_grad()

    outputs = model(X_train_tensor)
    loss = criterion(outputs, y_train_tensor)

    loss.backward()
    optimizer.step()

    train_losses.append(loss.item())

    # Validation
    model.eval()
    with torch.no_grad():
        test_outputs = model(X_test_tensor)
        test_loss = criterion(test_outputs, y_test_tensor)
        test_losses.append(test_loss.item())

    if epoch % 10 == 0:
        print(f"Epoch {epoch}, Train Loss: {loss.item():.4f}, Test Loss: {test_loss.item():.4f}")

Epoch 0, Train Loss: 2.7951, Test Loss: 2.7873
Epoch 10, Train Loss: 2.7872, Test Loss: 2.7772
Epoch 20, Train Loss: 2.7804, Test Loss: 2.7673
Epoch 30, Train Loss: 2.7747, Test Loss: 2.7586
Epoch 40, Train Loss: 2.7689, Test Loss: 2.7500


In [49]:
# Evaluate accuracy
from sklearn.metrics import accuracy_score

model.eval()
with torch.no_grad():
    preds = model(X_test_tensor)
    predicted = torch.argmax(preds, axis=1)

accuracy = accuracy_score(y_test_tensor, predicted)
print("Test Accuracy:", accuracy)

Test Accuracy: 0.14135988782470232


The model is underfitting because both training and test losses are high and similar, and the accuracy is very low (~14%), indicating that the model fails to learn meaningful patterns from the data.